add feature : allow kubernetes cluster and docker cluster to deploy and undeploy services

ykwd opened the pull request.

ykwd committed:
fix a bug about monitor's page of service

goksyli committed:
small fix

goksyli committed:
Merge branch 'master' of https://github.com/ykwd/rDSN

goksyli committed:
fix deployment scripts and add service url to cluster_scheduler

goksyli committed:
add deployment service unschedule

goksyli committed:
fix deployment service undeploy and multi-instance name conflict

goksyli committed:
add undeploy callback and fix some bugs

goksyli committed:
fix k8s deployment scripts and deployment service bugs

ykwd committed:
Merge remote-tracking branch 'upstream/master'

goksyli committed:
Merge branch 'master' of https://github.com/imzhenyu/rDSN

ykwd committed:
add machine pool manager

ykwd committed:
add sample config for machine pool manager

goksyli committed:
add pool manager to docker scheduler

goksyli committed:
merge from ykwd master

ykwd committed:
add config file 'nodes' for deployment service

goksyli committed:
Merge branch 'master' of https://github.com/ykwd/rDSN

goksyli committed:
fix bugs of scheduler provider

ykwd committed:
change some interfaces of machine_pool_mgr to integerate with cluster scheduler

goksyli committed:
add machine pool init

goksyli committed:
merge from ykwd master

goksyli committed:
fix bugs

ykwd committed:
fix some bugs :
1. fix a bug when long term operation such as deployment or undeployment block other cli call;
2. remove the use of a stale iterator in case of potential modification to the container;
3. add a feature so that monitor can display 'UNDEPLOYING' state

ykwd committed:
merge to upstream/master, fix a bug of upstream/master

imzhenyu commented on src/dist/cluster_scheduler/scheduler_providers/docker_scheduler.cpp:
```diff
@@ -160,6 +140,47 @@ error_code docker_scheduler::initialize()
+        for( auto& machine: a_list )
+        {
+            std::string command = "echo "+ machine +">> " + machine_file; 
+            system(command.c_str());
```
Should use std::ofstream etc. to make it more reliable. We should use `system` etc. only when we don't have other better approaches.

ykwd committed:
1. optimize : change the write file method from system call 'echo' to std::ofstream at docker cluster scheduler;
2. fix a bug : the remote directory is undefined when docker cluster undeploy a service;
3. fix a bug : a missing return value of function get_machines of machine_pool_mgr will cause core dump

ykwd committed:
Merge remote-tracking branch 'upstream/master'
